bittide/src/Bittide/DoubleBufferedRam.hs

```diff
@@ -108,24 +107,29 @@ wbStorageDP initial aM2S bM2S = (aS2M, bS2M)
     Dict -> wbStorage' @_ @depth @(Max awA awB) initial storageIn
 
   storageIn :: Signal dom (WishboneM2S (Max awA awB) 4 (Bytes 4))
-  storageIn = mux (nowSelected .==. pure A) (resizeM2SAddr <$> aM2S) (resizeM2SAddr <$> bM2S)
+  storageIn = mux (nowActive .==. pure A) (resizeM2SAddr <$> aM2S) (resizeM2SAddr <$> bM2S)
 
   -- We keep track of ongoing transactions to respect Read-modify-write operations.
-  lastSelected = register A nowSelected
-  nowSelected = selectNow <$> aM2S <*> bM2S <*> lastSelected
-  selectNow (busCycle -> cycA) (busCycle -> cycB) lastSel
-    | lastSel == B && cycB = B
-    | cycA && not cycB     = A
-    | otherwise            = A
-
-  (aS2M, bS2M) = unbundle $ mux (nowSelected .==. pure A)
+  nowActive = register A nextActive
+  nextActive = selectNow <$> nowActive <*> aM2S <*> aS2M <*> bM2S <*> bS2M
+  active WishboneM2S{busCycle, strobe} = busCycle && strobe
+  terminated WishboneS2M{acknowledge, err} = acknowledge || err
+  selectNow aorb am2s as2m bm2s bs2m  =
+    case (aorb, active am2s, terminated as2m, active bm2s, terminated bs2m) of
+      (_, True , _    , False, _     ) -> A
+      (_, False, _    , True , _     ) -> B
+      (A, True , True , True , _     ) -> B
+      (B, True , _    , True , True  ) -> A
+      _                                -> aorb
```

Comment:
There's still no tests for `wbStorageDP`, right? Maybe it makes sense to add some since it's changed here?

Reply:
Yes, but also: https://github.com/bittide/bittide-hardware/issues/304. We're going to push through because a lot of PRs depend on this one. Guess we'll see it falling apart later....